pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,154 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"strings"
+	"sync"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/jaeger"
+	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	otrace "go.opentelemetry.io/otel/trace"
+)
+
+func initJaegerExporter(config *CollectorConfig) (sdktrace.SpanExporter, error) {
+	return jaeger.New(jaeger.WithCollectorEndpoint(jaeger.WithEndpoint(
+		config.EndpointURL,
+	)))
+}
+
+func initWriterExporter(config *WriterConfig) (sdktrace.SpanExporter, error) {
+	var opts []stdouttrace.Option
+	if config.Pretty {
+		opts = append(opts, stdouttrace.WithPrettyPrint())
+	}
+	if !config.Timestamps {
+		opts = append(opts, stdouttrace.WithoutTimestamps())
+	}
+
+	var w io.Writer
+	switch strings.ToLower(config.Destination) {
+	case "stdout":
+		w = os.Stdout
+
+	case "stderr":
+		w = os.Stderr
+
+	default:
+		return nil, errors.New("unknown writer destination")
+	}
+	opts = append(opts, stdouttrace.WithWriter(w))
+
+	return stdouttrace.New(opts...)
+}
+
+func initResource(config *ResourceConfig) (*resource.Resource, error) {
+	rsc, err := resource.New(context.Background(),
+		resource.WithContainer(),
+		resource.WithProcess(),
+		resource.WithOS(),
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(config.ServiceName),
+			semconv.ServiceVersionKey.String(config.ServiceVersion),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fill empty values with defaults
+	return resource.Merge(resource.Default(), rsc)
+}
+
+var initOnce sync.Once
+
+// InitTelemetry initializes the telemetry package and returns the tracer provider.
+// If telemetry is not enabled it returns a noop tracer provider instead.
+func InitTelemetry(config *Config) (otrace.TracerProvider, error) {
+	if !config.Enable {
+		return otrace.NewNoopTracerProvider(), nil
+	}
+
+	var firstTime bool
```

Comment:
To be honest we're guaranteed to lose the race to set the global tracer provider before the other external packages, and that's fine. We can just avoid setting the global once and having this 'initialize only once' logic in place.